pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +102,310 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 		return fmt.Errorf("unhealthy members found during reconciling members")
 	}
 
-	// etcd is healthy, decide if we need to scale
-	podToAdd, err := c.getEtcdPodToAddToMembership(ctx)
-	switch {
-	case err != nil:
-		return fmt.Errorf("could not get etcd pod: %w", err)
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
+	// Add a learner member if next peer found
+	var errs []error
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("could not get etcd peerURL to add :%w", err))
+	}
+	if len(peerURL) > 0 {
+		err = c.etcdClient.MemberAddAsLearner(ctx, peerURL)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to add learner member :%w", err))
+		}
+	}
+
+	// Attempt to promote a learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		errs = append(errs, fmt.Errorf("failed to promote learner: %w", err))
+	}
+
+	return kerrors.NewAggregate(errs)
+}
+
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return "", err
+	}
+
+	nodesMapToVotingMembers, err := c.allNodesMapToVotingMembers(nodes)
```

Comment:
it could return a list of nodes that don't match

Reply:
So we only iterate over those nodes? Yeah I guess we could do that.